### PR TITLE
Add measured parameter descriptions config file for DSWx-S1 PGE

### PR DESCRIPTION
**Pull request by RKuttruff**
Probably still needs some work re: tests

## Description
- Can configure a YAML file to specify the descriptions of metadata attributes in generated ISO XML files
- Additional fix for handling of missing ISO XML template path in runconfig (it is not required but will raise a `KeyError` if accessed)

## Descriptions File Format
```yaml
METADATA_ATTRIBUTE_NAME:
  description: "Attribute description"
ANOTHER_METADATA_ATTRIBUTE_NAME:
  description: "Another attribute description"
  ...
```

## Affected Issues
- #499 

## Testing
- Adapted unit tests for runconfig to test for new parameter
- Existing unit tests for DSWx-S1 should handle insufficient descriptions file coverage

### CODE REVIEW

**Comment by RKuttruff** [comment]
Leaving this as draft for now so I can some more tests first

File: src/opera/pge/base/base_pge.py [hunk 1]
```diff
@@ -177,6 +179,39 @@ def _configure_logger(self):
             self.qa_logger.info(self.name, ErrorCode.LOG_FILE_INIT_COMPLETE,
                                 'Log file configuration complete')
 
+    def _validate_iso_descriptions(self):
+        """
+        If given, check if the run-config description file exists and is valid
+        """
```
Comment (JimHofman) [inline src/opera/pge/base/base_pge.py:185]: Just a 'nit' but single line docstrings are done on one line:   """<single line>"""

Comment (RKuttruff) [reply src/opera/pge/base/base_pge.py]: Fixed :)

File: src/opera/pge/base/schema/base_pge_schema.yaml [hunk 2]
```diff
@@ -29,7 +29,8 @@ RunConfig:
         ErrorCodeBase: int(required=True)
         SchemaPath: str(required=True)
         AlgorithmParametersSchemaPath: str(required=False)
-        IsoTemplatePath: str(required=False)
+        IsoTemplatePath: str(required=False)  # TODO: Why is this not required if all the PGEs fail if it doesn't exist/is undefined?
+        IsoMeasuredParameterDescriptions: str(required=False)
```
Comment (JimHofman) [inline src/opera/pge/base/schema/base_pge_schema.yaml:33]: Good question

File: src/opera/util/error_codes.py
Comment (JimHofman) [inline src/opera/util/error_codes.py:1]: FYI:  Scott mentioned we are moving away from adding new error codes, I think he may want one generic code or something.  I'm not sure how he will want to handle this case.

**Review by JimHofman — COMMENTED** [review]
Excellent job Riley.  I'm glad you understand the schema and jinja2 stuff.

**Comment by RKuttruff** [comment]
I tweaked this slightly to simplify the descriptions yaml file

File: src/opera/pge/dswx_s1/templates/dswx_s1_measured_parameters.yaml [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+MGRS_COLLECTION_ACTUAL_NUMBER_OF_BURSTS:
+  description: The number of actual RTC-S1 bursts that are collected within the MGRS tile collection ID.
+MGRS_COLLECTION_EXPECTED_NUMBER_OF_BURSTS:
+  description: The number of expected RTC-S1 bursts that are supposed to be collected within the MGRS tile collection ID.
+MGRS_COLLECTION_MISSING_NUMBER_OF_BURSTS:
+  description: The number of missing RTC-S1 bursts.
+MGRS_POL_MODE:
+  description: "The description for the polarizations of the collected bursts. 'MIX_DUAL_POL': ['HH', 'HV', 'VV', 'VH'], 'MIX_DUAL_H_SINGLE_V_POL': ['HH', 'HV', 'VV'], 'MIX_DUAL_V_SINGLE_H_POL': ['VV', 'VH', 'HH'], 'MIX_SINGLE_POL': ['HH', 'VV'], 'DV_POL': ['VV', 'VH'], 'SV_POL': ['VV'], 'DH_POL': ['HH', 'HV'], 'SH_POL': ['HH'],"
+INPUT_DEM_SOURCE:
+  description: Description of the input DEM
```
Comment (collinss-jpl) [inline src/opera/pge/dswx_s1/templates/dswx_s1_measured_parameters.yaml:20]: A nitpick, but I can see someone complaining about this in the future: lets be consistent on period usage for these descriptions and always include one even if it was left out of the text copied from the product spec

File: src/opera/pge/dswx_s1/dswx_s1_pge.py [hunk 4]
```diff
@@ -441,17 +443,28 @@ def _augment_measured_parameters(self, measured_parameters):
...
-            # TODO: This is hardcoded for now, but we should eventually try to guess this
+            # TODO: This is hardcoded for now, but we should eventually try to guess this or add it to the descriptions
+            #  YAML file
             attr_type = 'processingInformation'
 
-            # TODO: Filling this attribute is also on the back burner for now
-            attr_description = f'TBA: Description for {name}'
+            attr_description = descriptions.setdefault(name, dict(description=missing_description_value))['description']
 
             attr_name = (name.title()
```
Comment (collinss-jpl) [inline src/opera/pge/dswx_s1/dswx_s1_pge.py:469]: More a note to remind myself, but a good future ticket will be to figure out a good way to refactor this into a separate mapping dictionary that is PGE specific. The code can then just iterate over this dictionary, making the `replace` call for each key/value pair

**Review by collinss-jpl — APPROVED** [review]
This is a great step forward with this feature, and has given me a couple ideas on what to refine next, great work @RKuttruff!

File: src/opera/pge/base/base_pge.py [hunk 5]
```diff
@@ -177,6 +179,37 @@ def _configure_logger(self):
...
+            description_file = abspath(description_file)
+
+            if not exists(description_file):
+                msg = f'Could not load description file {description_file}, file does not exist.'
+                self.logger.critical(self.name, ErrorCode.ISO_METADATA_DESCRIPTIONS_CONFIG_NOT_FOUND, msg)
+
+            # Schema maps parameter names to maps which contain descriptions. Using this approach to facilitate adding
+            # any new fields as needed, eg attribute type, data type override, etc
+            schema = yamale.make_schema(content=dedent("""
+            map(include('parameter'), key=str())
```
[the remark this answers is not in the document]
Comment (RKuttruff) [reply src/opera/pge/base/base_pge.py:196]: That makes sense. I figured it was small and simple enough that we could just inline it but moving it out to a file eventually should be fine too.

File: src/opera/pge/dswx_s1/dswx_s1_pge.py (hunk 4 above)
Comment (RKuttruff) [reply src/opera/pge/dswx_s1/dswx_s1_pge.py:469]: Another idea I had was just leveraging the existing file to include fields that "override" our guesses and derivations in the event we know they'll likely not be fully correct

File: src/opera/pge/dswx_s1/templates/dswx_s1_measured_parameters.yaml (hunk 3 above)
Comment (RKuttruff) [reply src/opera/pge/dswx_s1/templates/dswx_s1_measured_parameters.yaml]: Done

File: src/opera/pge/dswx_s1/templates/OPERA_ISO_metadata_L3_DSWx_S1_template.xml.jinja2 [hunk 6]
```diff
@@ -602,7 +602,7 @@
                                         </eos:EOS_AdditionalAttributeDescription>
                                     </eos:reference>
                                     <eos:value>
-                                        <gco:CharacterString>{{ attribute.value }}{# ISO_OPERA_areaOrPoint #}</gco:CharacterString>
+                                        <gco:CharacterString>{{ attribute.value }}</gco:CharacterString>
                                     </eos:value>
                                 </eos:AdditionalAttribute>
                                 {%  endfor %}
```
[the remark this answers is not in the document]
Comment (RKuttruff) [reply src/opera/pge/dswx_s1/templates/OPERA_ISO_metadata_L3_DSWx_S1_template.xml.jinja2]: Done

**Pull request merged**